### PR TITLE
Add HuggingFace embedding consistent with cloud service

### PR DESCRIPTION
**Pull request by dpsimpson**
I've added HuggingFace models to the embeddings options in outlines.

Some things:

- There would be less code if we just used `sentence_transformers`, but this is consistent with what happens in `cloud` right now (a light re-write of that code. Ben's argument for doing manual pooling is to reduce dependency bloat.)
- It adds a `batch_size` parameter to the generated function, in the case where you want to do batching. The default processes the input all in one batch. Would it be better for the default to do batches of 1?
- I haven't tested it with a gpu, but it should be lowered properly if a gpu is available 
- I've written some tests that all pass. ~But for some reason, a lot of the other tests are failing on my machine.~ It turns out I was using Python 3.11, which explains all of this.
- It's been through the pre-commit circus.

lmk what you think needs changing etc.

### CODE REVIEW

File: outlines/models/hf_transformers.py [hunk 1]
```diff
@@ -421,3 +421,113 @@ def create_type_constraint(
...
+    -------
+    A function that will call run the HuggingFace embedding with the given parameters when
+    passed a prompt. It will use a GPU if there is one available.
+
+    """
+
+    def mean_pooling(model_output, attention_mask):
+        """Pools together the outputs from a sentence embedding model to generate a vector embedding.
+
+        Code modified from https://github.com/normal-computing/cloud/blob/main/embedding-services/main.py
```
Comment (brandonwillard) [inline outlines/models/hf_transformers.py:449]:
```suggestion
```

File: outlines/models/hf_transformers.py [hunk 2]
```diff
@@ -421,3 +421,113 @@ def create_type_constraint(
...
+        The batch size. If it is not provided, or if a negative value is given, the embeddings will be run as a single batch.
+
+
+        Returns
+        -------
+        A function that will call run the HuggingFace embedding with the given parameters when
+        passed a prompt. It will use a GPU if there is one available.
+
+        Code modified from https://github.com/normal-computing/cloud/blob/main/embedding-services/main.py
+
```
Comment (brandonwillard) [inline outlines/models/hf_transformers.py:481]:
```suggestion
```

File: outlines/models/hf_transformers.py [hunk 3]
```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
...
+        )
+
+    def get_embedding(
+        sentences: Union[List[str], str], *, batch_size: Optional[int] = None
+    ):
+        """Generates embeddings from mean-pooled sentence embeddings from HuggingFace sentence transformers.
+
+        Parameters
+        ----------
+        sentences: List[str] | str
```
Comment (rlouf) [inline outlines/models/hf_transformers.py:465]:
```suggestion
        sentences
```

File: outlines/models/hf_transformers.py [hunk 4]
```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
...
+            The strings to be embedded
+        batch_size: Optional[int]
```
Comment (rlouf) [inline outlines/models/hf_transformers.py:467]:
```suggestion
        batch_size
```

File: outlines/models/hf_transformers.py [hunk 5]
```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
...
+        The batch size. If it is not provided, or if a negative value is given, the embeddings will be run as a single batch.
```
Comment (rlouf) [inline outlines/models/hf_transformers.py:468]:
```suggestion
            The batch size. If it is not provided, or if a negative value is given, the embeddings will be run as a single batch.
```

File: outlines/models/hf_transformers.py [hunk 6]
```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
...
+        embeddings_list = []
+
+        for i in range(0, len(sentences), batch_size):
+            batch_sentences = sentences[i : i + batch_size]
+            encoded_input = tokenizer(
+                batch_sentences,
+                padding=True,
+                truncation=True,
+                return_tensors="pt",
+            )
```
Comment (rlouf) [inline outlines/models/hf_transformers.py:507]: What is the point of batching? Memory constraints?

File: outlines/models/hf_transformers.py [hunk 7]
```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
...
+                model_output, encoded_input["attention_mask"]
+            )
+            batch_embeddings = torch.nn.functional.normalize(
+                batch_embeddings, p=2, dim=1
+            )
+
+            embeddings_list.append(batch_embeddings)
+
+        embeddings_tensor = torch.cat(embeddings_list, dim=0)
+        return embeddings_tensor.cpu().numpy()[:n_sentences,]  # check slicing here
```
Comment (rlouf) [inline outlines/models/hf_transformers.py:525]: We should write a test to check the slicing. (what would be after `n_sentences`?)

File: outlines/models/hf_transformers.py [hunk 8]
```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
...
+
+
+        Returns
+        -------
+        A function that will call run the HuggingFace embedding with the given parameters when
+        passed a prompt. It will use a GPU if there is one available.
+
+        """
+        import torch
+        from transformers import AutoModel, AutoTokenizer
```
Comment (rlouf) [inline outlines/models/hf_transformers.py:478]: Is there any kind of check on `model_name` that we can enforce? For text generation this is done on the `transformers` size by calling `AutoModelForCausalLM`. Is there a similar model class in `transformers` that checks that the model is appropriate for the given task?

Maybe even something as simple as making sure that the model name starts with `sentence-transfomers/`.

File: tests/models/test_hf_transformers.py [hunk 9]
```diff
@@ -76,3 +81,27 @@ def test_is_in_multiple_samples():
...
+    model = HuggingFaceEmbeddings(EMBED_MODEL)
+    out = model(["This is a test sentence", "Another test sentence"])
+    assert isinstance(out, np.ndarray)
+    assert out.shape[0] == 2
+    assert out.shape[1] == 384
+
+
+def test_batch_embeddings():
+    model = HuggingFaceEmbeddings(EMBED_MODEL)
+    out = model(["This is a test sentence", "Another test sentence"], batch_size=4)
```
[the remark this answers is not in the document]
Comment (dpsimpson) [reply tests/models/test_hf_transformers.py:104]: It’s checking that the batching returns the right shape even after padding.

File: outlines/models/hf_transformers.py (hunk 6 above)
Comment (dpsimpson) [reply outlines/models/hf_transformers.py:507]: Also parallelism. Faster to reduce number of sends to gpu

**Comment by brosand** [comment]
> > * There would be less code if we just used `sentence_transformers`, but this is consistent with what happens in `cloud` right now (a light re-write of that code. Ben's argument for doing manual pooling is to reduce dependency bloat.)
> 
> Not an expert, but it looks like the `sentence-transformer` repos contains much more code than just a pooling function. With what models is the code in this PR compatible?

This code is specially built for the sgpt implementation. I would actually vote for moving to `sentence-transformers in the long-term, so I think it's reasonable to do it here (assuming there aren't any reproducibility problems

> > * It adds a `batch_size` parameter to the generated function, in the case where you want to do batching. The default processes the input all in one batch. Would it be better for the default to do batches of 1?
> 
> Has Ben found that chunking batches was necessary? At what scale?

Batch size really depends on hardware and the model that is being used. I think it's probably fine to set a default batch size of 128 or larger since most embedding models are pretty small, but I wouldn't really want to commit to any numbers without more information.
